src/compute/src/memory_management/memory_manager.rs

```diff
@@ -120,38 +142,50 @@ impl GlobalMemoryManager {
                 tracing::warn!("Jemalloc epoch advance failed! {:?}", e);
             }
 
-            let cur_total_bytes_used = jemalloc_allocated_mib.read().unwrap_or_else(|e| {
+            let jemalloc_allocated_mib = jemalloc_allocated_mib.read().unwrap_or_else(|e| {
                 tracing::warn!("Jemalloc read allocated failed! {:?}", e);
-                last_total_bytes_used
+                last_jemalloc_allocated_mib
             });
+            last_jemalloc_allocated_mib = jemalloc_allocated_mib;
 
-            // The strategy works as follow:
+            // ## Batch memory control
             //
-            // 1. When the memory usage is below the graceful threshold, we do not evict any caches
-            // and reset the step to 0.
+            // When the batch memory usage exceeds the threshold, we choose the query that uses the
+            // most memory and kills it.
+
+            let batch_used_memory_bytes = batch_manager.total_mem_usage();
+            if batch_used_memory_bytes > batch_memory_threshold {
+                batch_manager.kill_queries("excessive batch memory usage".to_string());
+            }
```

Comment:
The current interval of running the policy is 1 second, which was deliberately aligned to the default barrier interval, may be too long to catch all problematic batch queries.

The reasons of that are: 1) The total available memory of batch queries is smaller than streaming, so technically it may take less time to fill it up; 2) Batch queries are not limited to external source's throughput, so RisingWave always tried its best to read and process data.

May need some futher disccusion and fix in another PR.

Reply:
I'll add a switch on batch memory control that currently turns off by default.

Reply:
> 3. The interval between two runs.

-1 for this. The streaming policy actually depends on interval to decide the steps. Adjusting interval to other values than barrier_interval will cause unexpected behaviors.

Reply:
> -1 for this. The streaming policy actually depends on interval to decide the steps. Adjusting interval to other values than barrier_interval will cause unexpected behaviors.

Oh, in fact I mean the interval for checking batch memory usage.